### PR TITLE
Upstream some utils from Agora

### PR DESCRIPTION
**Pull request by chfanghr**
This pr is based off #57.

### CODE REVIEW

File: src/Plutarch/Orphans.hs [hunk 1]
```diff
@@ -1,17 +1,64 @@
...
 {-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE UndecidableInstances #-}
 {-# OPTIONS_GHC -Wno-orphans #-}
 
 module Plutarch.Orphans () where
 
 import Control.Composition (on, (.*))
+import Plutarch.Api.V2 (PDatumHash (..), PScriptHash (..))
```
Comment (SeungheonOh) [inline src/Plutarch/Orphans.hs:12]: This import wildcards should be spelled out.

File: src/Plutarch/Orphans.hs [hunk 2]
```diff
@@ -1,17 +1,64 @@
...
+    type PTryFromExcess PData (PAsData PDatumHash) = Flip Term PDatumHash
+    ptryFrom' opq = runTermCont $ do
+        unwrapped <- pfromData . fst <$> ptryFromC @(PAsData PByteString) opq
+
+        tcont $ \f ->
+            pif
+                -- Blake2b_256 hash: 256 bits/32 bytes.
+                (plengthBS # unwrapped #== 32)
+                (f ())
+                (ptraceError "ptryFrom(PDatumHash): must be 32 bytes long")
```
Comment (SeungheonOh) [inline src/Plutarch/Orphans.hs:44]: Shouldn't this be "... must be 28 bytes long"?

File: src/Plutarch/Extra/Tuple.hs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+module Plutarch.Extra.Tuple (pfstTuple, psndTuple) where
+
+import Plutarch.Api.V1.Tuple (PTuple)
+
+{- | Extract the first component of a 'PTuple'.
+
+     @since 3.0.2
+-}
+pfstTuple :: forall a b s. (PIsData a) => Term s (PTuple a b :--> a)
```
Comment (SeungheonOh) [inline src/Plutarch/Extra/Tuple.hs:11]: Kind signatures

File: src/Plutarch/Extra/Tuple.hs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+pfstTuple = phoistAcyclic $ plam $ pfromData . (pfield @"_0" #)
+
+{- | Extract the second component of a 'PTuple'.
+
+     @since 3.0.2
+-}
+psndTuple :: forall b a s. (PIsData b) => Term s (PTuple a b :--> b)
```
Comment (SeungheonOh) [inline src/Plutarch/Extra/Tuple.hs:18]: Kind signatures

File: src/Plutarch/Extra/ScriptContext.hs [hunk 5]
```diff
@@ -196,14 +220,101 @@ pfindDatum = phoistAcyclic $
...
+    phoistAcyclic $
+        plam $
+            pfromJust #.* ptryFromOutputDatum
+
+{- | Extract the datum hash from a 'POutputDatum', throw an error if the given
+     'POuptutDatum' doesn't contain a datum hash.
+
+     @since 3.0.2
+-}
+pfromDatumHash :: forall s. Term s (POutputDatum :--> PDatumHash)
```
Comment (SeungheonOh) [inline src/Plutarch/Extra/ScriptContext.hs:300]: Kind signatures

File: src/Plutarch/Extra/ScriptContext.hs [hunk 6]
```diff
@@ -196,14 +220,101 @@ pfindDatum = phoistAcyclic $
...
+        flip pmatch $ \case
+            POutputDatumHash ((pfield @"datumHash" #) -> hash) -> hash
+            _ -> ptraceError "not a datum hash"
+
+{- | Extract the inline datum from a 'POutputDatum', throw an error if the given
+     'POuptutDatum' is not an inline datum.
+
+     @since 3.0.2
+-}
+pfromInlineDatum :: forall s. Term s (POutputDatum :--> PDatum)
```
Comment (SeungheonOh) [inline src/Plutarch/Extra/ScriptContext.hs:312]: Kind signatures

**Review by SeungheonOh — CHANGES_REQUESTED** [review]

File: src/Plutarch/Orphans.hs (hunk 2 above)
Comment (chfanghr) [reply src/Plutarch/Orphans.hs:44]: It's a blake2b_256 hash, isn't it?

Comment (SeungheonOh) [reply src/Plutarch/Orphans.hs:44]: Oh, sorry I meant to commit this on L62.

File: src/Plutarch/Extra/Function.hs [hunk 7]
```diff
@@ -64,3 +67,49 @@ pflip ::
         s
         ((a :--> b :--> c) :--> b :--> a :--> c)
 pflip = phoistAcyclic $ plam $ \f y x -> f # x # y
+
+infixr 8 #.*
+infixr 8 #.**
+infixr 8 #.***
```
Comment (kozross) [inline src/Plutarch/Extra/Function.hs:73]: I find it more readable to put fixities after individual definitions, but this is a minor issue.

**Review by kozross — APPROVED** [review]

File: src/Plutarch/Extra/Function.hs (hunk 7 above)
Comment (chfanghr) [reply src/Plutarch/Extra/Function.hs:73]: Will fix!

**Review by peter-mlabs — APPROVED** [review]
LGTM!

**Review by SeungheonOh — APPROVED** [review]
LGTM. I hope there's a way to get rid of those overlapping instances but we are good to go at this point.

**Pull request merged**